File: android/src/main/kotlin/com/acs_plugin/calling/redux/middleware/CallingMiddleware.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.acs_plugin.calling.redux.middleware

import com.acs_plugin.calling.logger.Logger
import com.acs_plugin.calling.redux.Dispatch
import com.acs_plugin.calling.redux.Middleware
import com.acs_plugin.calling.redux.Store
import com.acs_plugin.calling.redux.action.Action
import com.acs_plugin.calling.redux.action.AudioSessionAction
import com.acs_plugin.calling.redux.action.CallDiagnosticsAction
import com.acs_plugin.calling.redux.action.CallingAction
import com.acs_plugin.calling.redux.action.CaptionsAction
import com.acs_plugin.calling.redux.action.ErrorAction
import com.acs_plugin.calling.redux.action.LifecycleAction
import com.acs_plugin.calling.redux.action.LocalParticipantAction
import com.acs_plugin.calling.redux.action.ParticipantAction
import com.acs_plugin.calling.redux.action.PermissionAction
import com.acs_plugin.calling.redux.action.RttAction
import com.acs_plugin.calling.redux.action.ToastNotificationAction
import com.acs_plugin.calling.redux.middleware.handler.CallingMiddlewareActionHandler
import com.acs_plugin.calling.redux.state.ReduxState

internal interface CallingMiddleware

internal class CallingMiddlewareImpl(
    private val callingMiddlewareActionHandler: CallingMiddlewareActionHandler,
    private val logger: Logger,
) :
    Middleware<ReduxState>,
    CallingMiddleware {

    override fun invoke(store: Store<ReduxState>) = { next: Dispatch ->
        { action: Action ->
            logger.info(action.toString())
            when (action) {
                is LifecycleAction.EnterBackgroundTriggered -> {
                    callingMiddlewareActionHandler.enterBackground(store)
                }
                is LifecycleAction.EnterForegroundTriggered -> {
                    callingMiddlewareActionHandler.enterForeground(store)
                }
                is LocalParticipantAction.CameraPreviewOnRequested -> {
                    callingMiddlewareActionHandler.requestCameraPreviewOn(store)
                }
                is LocalParticipantAction.CameraPreviewOnTriggered -> {
                    callingMiddlewareActionHandler.turnCameraPreviewOn(store)
                }
                is LocalParticipantAction.CameraOffTriggered -> {
                    callingMiddlewareActionHandler.turnCameraOff(store)
                }
                is LocalParticipantAction.CameraOnRequested -> {
                    callingMiddlewareActionHandler.requestCameraOn(store)
                }
                is LocalParticipantAction.CameraOnTriggered -> {
                    callingMiddlewareActionHandler.turnCameraOn(store)
                }
                is LocalParticipantAction.CameraSwitchTriggered -> {
                    callingMiddlewareActionHandler.switchCamera(store)
                }
                is LocalParticipantAction.MicOffTriggered -> {
                    callingMiddlewareActionHandler.turnMicOff(store)
                }
                is AudioSessionAction.AudioFocusApproved -> {
                    store.dispatch(CallingAction.ResumeRequested())
                }
                is AudioSessionAction.AudioFocusInterrupted -> {
                    store.dispatch(CallingAction.HoldRequested())
                }
                is LocalParticipantAction.MicOnTriggered -> {
                    callingMiddlewareActionHandler.turnMicOn(store)
                }
                is LocalParticipantAction.AudioStateOperationUpdated -> {
                    callingMiddlewareActionHandler.onUpdateAudioStateOperation(action.audioOperationalStatus, store)
                }
                is LocalParticipantAction.BlurPreviewOnTriggered -> {
                    callingMiddlewareActionHandler.turnBlurOn(store)
                }
                is LocalParticipantAction.BlurPreviewOffTriggered -> {
                    callingMiddlewareActionHandler.turnBlurOff(store)
                }
                is LocalParticipantAction.NoiseSuppressionOnTriggered -> {
                    callingMiddlewareActionHandler.turnNoiseSuppressionOn(store)
                }
                is LocalParticipantAction.NoiseSuppressionOffTriggered -> {
                    callingMiddlewareActionHandler.turnNoiseSuppressionOff(store)
                }
                is LocalParticipantAction.RaiseHandTriggered -> {
                    callingMiddlewareActionHandler.raiseHand(store)
                }
                is LocalParticipantAction.LowerHandTriggered -> {
                    callingMiddlewareActionHandler.lowerHand(store)
                }
                is LocalParticipantAction.SendReactionTriggered -> {
                    callingMiddlewareActionHandler.sendReaction(action.reactionType, store)
                }
                is LocalParticipantAction.ShareScreenTriggered -> {
                    callingMiddlewareActionHandler.turnShareScreenOn(store)
                }
                is LocalParticipantAction.StopShareScreenTriggered -> {
                    callingMiddlewareActionHandler.turnShareScreenOff(store)
                }
                is CallingAction.HoldRequested -> {
                    callingMiddlewareActionHandler.hold(store)
                }
                is CallingAction.ResumeRequested -> {
                    callingMiddlewareActionHandler.resume(store)
                }
                is CallingAction.CallEndRequested -> {
                    callingMiddlewareActionHandler.endCall(store)
                }
                is CallingAction.SetupCall -> {
                    callingMiddlewareActionHandler.setupCall(store)
                }
                is CallingAction.CallStartRequested -> {
                    callingMiddlewareActionHandler.startCall(store)
                }
                is CallingAction.CallRequestedWithoutSetup -> {
                    callingMiddlewareActionHandler.callSetupWithSkipSetupScreen(store)
                }
                is PermissionAction.CameraPermissionIsSet -> {
                    callingMiddlewareActionHandler.onCameraPermissionIsSet(store)
                }
                is ErrorAction.EmergencyExit -> {
                    callingMiddlewareActionHandler.exit(store)
                }
                is ParticipantAction.AdmitAll -> {
                    callingMiddlewareActionHandler.admitAll(store)
                }
                is ParticipantAction.Admit -> {
                    callingMiddlewareActionHandler.admit(action.userIdentifier, store)
                }
                is ParticipantAction.Reject -> {
                    callingMiddlewareActionHandler.reject(action.userIdentifier, store)
                }
                is ParticipantAction.Remove -> {
                    callingMiddlewareActionHandler.removeParticipant(action.userIdentifier, store)
                }
                is LocalParticipantAction.SetCapabilities -> {
                    callingMiddlewareActionHandler.setCapabilities(action.capabilities, store)
                }
                is CallDiagnosticsAction.NetworkQualityCallDiagnosticsUpdated -> {
                    callingMiddlewareActionHandler
                        .onNetworkQualityCallDiagnosticsUpdated(action.networkQualityCallDiagnosticModel, store)
                }
                is CallDiagnosticsAction.NetworkCallDiagnosticsUpdated -> {
                    callingMiddlewareActionHandler
                        .onNetworkCallDiagnosticsUpdated(action.networkCallDiagnosticModel, store)
                }
                is CallDiagnosticsAction.MediaCallDiagnosticsUpdated -> {
                    callingMiddlewareActionHandler
                        .onMediaCallDiagnosticsUpdated(action.mediaCallDiagnosticModel, store)
                }
                is ToastNotificationAction.DismissNotification -> {
                    callingMiddlewareActionHandler.dismissNotification(store)
                }
                is LocalParticipantAction.AudioDeviceChangeRequested -> {
                    callingMiddlewareActionHandler.onAudioDeviceChangeRequested(action.requestedAudioDevice, store)
                }
                is LocalParticipantAction.AudioDeviceChangeSucceeded -> {
                    callingMiddlewareActionHandler.onAudioDeviceChangeSucceeded(action.selectedAudioDevice, store)
                }
                is AudioSessionAction.AudioFocusRequesting -> {
                    callingMiddlewareActionHandler.onAudioFocusRequesting(store)
                }
                is CaptionsAction.StartRequested -> {
                    callingMiddlewareActionHandler.startCaptions(action.language, store)
                }
                is CaptionsAction.StopRequested -> {
                    callingMiddlewareActionHandler.stopCaptions(store)
                }
                is CaptionsAction.SetSpokenLanguageRequested -> {
                    callingMiddlewareActionHandler.setCaptionsSpokenLanguage(action.language, store)
                }
                is CaptionsAction.SetCaptionLanguageRequested -> {
                    callingMiddlewareActionHandler.setCaptionsCaptionLanguage(action.language, store)
                }
                is RttAction.SendRtt -> {
                    callingMiddlewareActionHandler.sendRttMessage(action.message, action.isFinalized, store)
                }
            }
            next(action)
        }
    }
}

File: ios/Classes/AzureCommunicationUICalling/AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Redux/Middleware/CallingMiddleware.swift
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//  Licensed under the MIT License.
//

import Combine

extension Middleware {
    static func liveCallingMiddleware(callingMiddlewareHandler actionHandler: CallingMiddlewareHandling)
    -> Middleware<AppState, acs_plugin.Action> {
        .init(
            apply: { dispatch, getState in
                return { next in
                    return { action in
                        switch action {
                        case .callingAction(let callingAction):
                            handleCallingAction(callingAction, actionHandler, getState, dispatch)

                        case .localUserAction(let localUserAction):
                            handleLocalUserAction(localUserAction, actionHandler, getState, dispatch)

                        case .permissionAction(let permissionAction):
                            handlePermissionAction(permissionAction, actionHandler, getState, dispatch)

                        case .lifecycleAction(let lifecycleAction):
                            handleLifecycleAction(lifecycleAction, actionHandler, getState, dispatch)

                        case .audioSessionAction(let audioAction):
                            handleAudioSessionAction(audioAction, actionHandler, getState, dispatch)
                        case .remoteParticipantsAction(let action):
                            handleRemoteParticipantAction(action, actionHandler, getState, dispatch)
                        case .captionsAction(let action):
                            handleCaptionsAction(action, actionHandler, getState, dispatch)
                        case .errorAction,
                                .compositeExitAction,
                                .callingViewLaunched:
                            break
                        case .callDiagnosticAction(let action):
                            handleCallDiagnisticAction(action, actionHandler, getState, dispatch)
                        case .toastNotificationAction(let action):
                            handleToastNotificationAction(action, actionHandler, getState, dispatch)
                        default:
                            break
                        }
                        return next(action)
                    }
                }
            }
        )
    }
}

private func handleCallingAction(_ action: CallingAction,
                                 _ actionHandler: CallingMiddlewareHandling,
                                 _ getState: () -> AppState,
                                 _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .setupCall:
        actionHandler.setupCall(state: getState(), dispatch: dispatch)
    case .callStartRequested:
        actionHandler.startCall(state: getState(), dispatch: dispatch)
    case .callEndRequested:
        actionHandler.endCall(state: getState(), dispatch: dispatch)
    case .holdRequested:
        actionHandler.holdCall(state: getState(), dispatch: dispatch)
    case .resumeRequested:
        actionHandler.resumeCall(state: getState(), dispatch: dispatch)
    case .recordingStateUpdated(let isRecordingActive):
        actionHandler.recordingStateUpdated(state: getState(),
                                            dispatch: dispatch,
                                            isRecordingActive: isRecordingActive)
    case .transcriptionStateUpdated(let isTranscriptionActive):
        actionHandler.transcriptionStateUpdated(state: getState(),
                                                dispatch: dispatch,
                                                isTranscriptionActive: isTranscriptionActive)
    default:
        break
    }
}

private func handleLocalUserAction(_ action: LocalUserAction,
                                   _ actionHandler: CallingMiddlewareHandling,
                                   _ getState: () -> AppState,
                                   _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .cameraPreviewOnTriggered:
        actionHandler.requestCameraPreviewOn(state: getState(), dispatch: dispatch)
    case .cameraOnTriggered:
        actionHandler.requestCameraOn(state: getState(), dispatch: dispatch)
    case .cameraOffTriggered:
        actionHandler.requestCameraOff(state: getState(), dispatch: dispatch)
    case .cameraSwitchTriggered:
        actionHandler.requestCameraSwitch(state: getState(), dispatch: dispatch)
    case .microphoneOffTriggered:
        actionHandler.requestMicrophoneMute(state: getState(), dispatch: dispatch)
    case .microphoneOnTriggered:
        actionHandler.requestMicrophoneUnmute(state: getState(), dispatch: dispatch)
    case .setCapabilities(let capabilities):
        actionHandler.setCapabilities(capabilities: capabilities, state: getState(), dispatch: dispatch)
    case .onCapabilitiesChanged(let event):
        actionHandler.onCapabilitiesChanged(event: event, state: getState(), dispatch: dispatch)
    case .raiseHandRequested:
        actionHandler.requestRaiseHand(state: getState(), dispatch: dispatch)
    case .sendReaction(let reaction):
        actionHandler.sendReaction(reaction, state: getState(), dispatch: dispatch)
    case .lowerHandRequested:
        actionHandler.requestLowerHand(state: getState(), dispatch: dispatch)
    case .backgroundEffectRequested(let effect):
        actionHandler.requestBackgroundEffect(effect: effect, state: getState(), dispatch: dispatch)
    case .audioDeviceChangeSucceeded:
        if (getState().localUserState.incomingAudioState.operation == .muted) {
            actionHandler.unMuteCall(state: getState(), dispatch: dispatch)
        }
    case .muteIncomingAudioRequested:
        actionHandler.muteCall(state: getState(), dispatch: dispatch)
    case .noiseSuppressionCallOn:
        actionHandler.noiseSuppressionCallOn(state: getState(), dispatch: dispatch)
    case .noiseSuppressionCallOff:
        actionHandler.noiseSuppressionCallOff(state: getState(), dispatch: dispatch)
    case .screenShareOnRequested:
        actionHandler.requestScreenSharingStream(state: getState(), dispatch: dispatch)
    case .screenShareOffRequested:
        actionHandler.requestStopScreenSharingStream(state: getState(), dispatch: dispatch)
    case .showChat(let azureCorrelationId):
        actionHandler.showChat(azureCorrelationId: azureCorrelationId)
        
    case .cameraOnSucceeded,
            .cameraOnFailed,
            .cameraOffSucceeded,
            .cameraOffFailed,
            .cameraPausedSucceeded,
            .cameraPausedFailed,
            .cameraSwitchSucceeded,
            .cameraSwitchFailed,
            .microphoneOnFailed,
            .microphoneOffFailed,
            .microphoneMuteStateUpdated,
            .microphonePreviewOn,
            .microphonePreviewOff,
            .noiseSuppressionPreviewOn,
            .noiseSuppressionPreviewOff,
            .audioDeviceChangeRequested,
            .audioDeviceChangeFailed,
            .changeDisplayNameRequested,
            .gridLayoutSelected,
            .speakerLayoutSelected,
            .raiseHandFailed,
            .lowerHandFailed,
            .raiseHandSucceeded,
            .lowerHandSucceeded,
            .backgroundEffectSetFailed,
            .screenShareOnTriggeredFailed,
            .screenShareOffTriggeredFailed,
            .resetLocalUserReaction,
            .screenShareOnSucceeded,
            .screenShareOffSucceeded,
            .participantRoleChanged:
        break
    }
}

private func handlePermissionAction(_ action: PermissionAction,
                                    _ actionHandler: CallingMiddlewareHandling,
                                    _ getState: () -> AppState,
                                    _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .cameraPermissionGranted:
        actionHandler.onCameraPermissionIsSet(state: getState(), dispatch: dispatch)
    case .audioPermissionGranted:
        actionHandler.onMicPermissionIsGranted(state: getState(), dispatch: dispatch)
    case .audioPermissionRequested,
            .audioPermissionDenied,
            .audioPermissionNotAsked,
            .cameraPermissionRequested,
            .cameraPermissionDenied,
            .cameraPermissionNotAsked:
        break
    }
}

private func handleLifecycleAction(_ action: LifecycleAction,
                                   _ actionHandler: CallingMiddlewareHandling,
                                   _ getState: () -> AppState,
                                   _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .backgroundEntered:
        actionHandler.enterBackground(state: getState(), dispatch: dispatch)
    case .foregroundEntered:
        actionHandler.enterForeground(state: getState(), dispatch: dispatch)
    case .willTerminate:
        if getState().callingState.status == .connected {
            actionHandler.endCall(state: getState(), dispatch: dispatch)
        }
    }
}

private func handleAudioSessionAction(_ action: AudioSessionAction,
                                      _ actionHandler: CallingMiddlewareHandling,
                                      _ getState: () -> AppState,
                                      _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .audioInterrupted:
        actionHandler.audioSessionInterrupted(state: getState(), dispatch: dispatch)
    case .audioInterruptEnded,
            .audioEngaged:
        break
    }
}

private func handleRemoteParticipantAction(_ action: RemoteParticipantsAction,
                                           _ actionHandler: CallingMiddlewareHandling,
                                           _ getState: () -> AppState,
                                           _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .admitAll:
        actionHandler.admitAllLobbyParticipants(state: getState(), dispatch: dispatch)
    case .declineAll:
        actionHandler.declineAllLobbyParticipants(state: getState(), dispatch: dispatch)
    case .admit(participantId: let participantId):
        actionHandler.admitLobbyParticipant(state: getState(), dispatch: dispatch, participantId: participantId)
    case .decline(participantId: let participantId):
        actionHandler.declineLobbyParticipant(state: getState(), dispatch: dispatch, participantId: participantId)
    case .remove(participantId: let participantId):
        actionHandler.removeParticipant(state: getState(), dispatch: dispatch, participantId: participantId)
    default:
        break
    }
}

private func handleCallDiagnisticAction(_ action: DiagnosticsAction,
                                        _ actionHandler: CallingMiddlewareHandling,
                                        _ getState: () -> AppState,
                                        _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .networkQuality(diagnostic: let diagnostic):
        actionHandler.onNetworkQualityCallDiagnosticsUpdated(
            state: getState(), dispatch: dispatch, diagnisticModel: diagnostic)
    case .network(diagnostic: let diagnostic):
        actionHandler.onNetworkCallDiagnosticsUpdated(
            state: getState(), dispatch: dispatch, diagnisticModel: diagnostic)
    case .media(diagnostic: let diagnostic):
        actionHandler.onMediaCallDiagnosticsUpdated(state: getState(), dispatch: dispatch, diagnisticModel: diagnostic)
    default:
        break
    }
}

private func handleToastNotificationAction(_ action: ToastNotificationAction,
                                           _ actionHandler: CallingMiddlewareHandling,
                                           _ getState: () -> AppState,
                                           _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .dismissNotification:
        actionHandler.dismissNotification(state: getState(), dispatch: dispatch)
    default:
        break
    }
}

private func handleCaptionsAction(_ action: CaptionsAction,
                                  _ actionHandler: CallingMiddlewareHandling,
                                  _ getState: () -> AppState,
                                  _ dispatch: @escaping ActionDispatch) {
    switch action {
    case .turnOnCaptions(language: let language):
        actionHandler.startCaptions(state: getState(), dispatch: dispatch, language: language)
    case .turnOffCaptions:
        actionHandler.stopCaptions(state: getState(), dispatch: dispatch)
    case .setSpokenLanguageRequested(language: let language):
        actionHandler.setCaptionsSpokenLanguage(state: getState(), dispatch: dispatch, language: language)
    case .setCaptionLanguageRequested(language: let language):
        actionHandler.setCaptionsLanguage(state: getState(), dispatch: dispatch, language: language)
    default:
        break
    }
}
